File: .config/nitrogen/downloadRedditWallpaper.js
const https = require('https');
const fs = require('fs');
const path = require('path');
const url = require('url');

const randArrItem = (arr) => arr[Math.floor(Math.random() * arr.length)];

const subreddits = ['wallpaper', 'earthporn', 'spaceporn'];

const redditUrl = `https://www.reddit.com/r/${randArrItem(subreddits)}/top.json?sort=top&t=month&limit=100`;
const parsedRedditUrl = url.parse(redditUrl);
const getOptions = {
  host: parsedRedditUrl.host,
  path: parsedRedditUrl.path,
  headers: {
    'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.134 Safari/537.36'
  }
};

const isValidPost = (post) => {
    const url = post?.data?.url_overridden_by_dest ?? "";
    return /\.(jpg|png|gif)$/i.test(url);
};

https.get(getOptions, (res) => {
  let data = '';
  res.on('data', (chunk) => {
    data += chunk;
  });
  res.on('end', () => {
    const json = JSON.parse(data);
    const posts = json.data.children;
    const postsWithImages = posts.filter(isValidPost);
    const randomPost = randArrItem(postsWithImages);
    const imageUrl = randomPost.data.url_overridden_by_dest;

    const imagePath = '/home/kyle/.config/nitrogen/bg.png';

    https.get(imageUrl, (res) => {
      res.pipe(fs.createWriteStream(imagePath));
      // console.log(`Image saved to ${imagePath}`);
    });
  });
}).on('error', (err) => {
  console.error(err);
});

/*
#!/bin/bash

/usr/local/bin/node /home/kyle/.config/nitrogen/downloadRedditWallpaper.js

export DISPLAY=":0"
nitrogen --restore
*/

/*
kyle@x260:~$ cat .config/nitrogen/bg-saved.cfg
[xin_-1]
file=/home/kyle/.config/nitrogen/bg.png
mode=5
bgcolor=#000000
*/

File: .config/nitrogen/downloadRedditWallpaper.py
#!/bin/python3

import random
import json
from urllib.request import Request, urlopen

image_save_path = '/home/kyle/.config/nitrogen/bg.png'
subreddits = ['wallpaper', 'earthporn', 'spaceporn']

reddit_url = f"https://www.reddit.com/r/{random.choice(subreddits)}/top.json?sort=top&t=month&limit=100"
headers = {
    'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.134 Safari/537.36'
}

def http_get(url):
    return urlopen(Request(url, headers=headers))

def is_valid_post(post):
    url = post.get('data', {}).get('url_overridden_by_dest', "")
    return bool(url) and url.lower().endswith(('.jpg', '.png', '.gif'))

response = http_get(reddit_url)

if response.status == 200:
    data = response.read().decode('utf-8')
    json_data = json.loads(data)
    posts = json_data['data']['children']
    posts_with_images = [post for post in posts if is_valid_post(post)]
    random_post = random.choice(posts_with_images)
    image_url = random_post['data']['url_overridden_by_dest']

    image_response = http_get(image_url)

    if image_response.status == 200:
        image_data = image_response.read()
        with open(image_save_path, 'wb') as image_file:
            image_file.write(image_data)
    else:
        print(f"Failed to fetch image from {image_url}")
else:
    print(f"Failed to fetch posts from {reddit_url}")

'''
kyle@x260:~$ cat .config/nitrogen/reroll.sh
#!/bin/bash

/home/kyle/.config/nitrogen/downloadRedditWallpaper.py

export DISPLAY=":0"
nitrogen --restore
'''

'''
kyle@x260:~$ cat .config/nitrogen/bg-saved.cfg
[xin_-1]
file=/home/kyle/.config/nitrogen/bg.png
mode=5
bgcolor=#000000
'''

